src/kakarot/precompiles/datacopy.cairo

```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: MIT
+
+%lang starknet
+
+// Starkware dependencies
+from starkware.cairo.common.cairo_builtins import HashBuiltin, BitwiseBuiltin
+from starkware.cairo.common.memcpy import memcpy
+
+// Internal dependencies
+from kakarot.model import model
+from utils.utils import Helpers
+from kakarot.memory import Memory
+from kakarot.execution_context import ExecutionContext
+
+// @title DataCopy precompile
+// @custom:precompile
+// @custom:address 0x04
+// @notice This precompile serves as a cheaper way to copy data in memory
+// @author @abdelhamidbakhta
+// @custom:namespace PrecompileDataCopy
+namespace PrecompileDataCopy {
+    const PRECOMPILE_ADDRESS = 0x04;
+    const GAS_COST_DATACOPY = 15;
+
+    // @notice Run the precompile.
+    // @param ctx The pointer to the execution context
+    // @return The pointer to the updated execution context.
+    func run{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
+        alloc_locals;
+
+        let (minimum_word_size) = Helpers.minimum_word_count(ctx.call_context.calldata_len);
+        let (output, output_len) = data_copy(
+            input=ctx.call_context.calldata, input_len=ctx.call_context.calldata_len
+        );
+
+        // compensating for `ret_offset` param in call opcodes
+        // currently stored at 0th place in `return_data`,
+        // see system_operations::CallHelpers.prepare_args
+        memcpy(ctx.return_data + ctx.return_data_len - 1, output, output_len);
```

Comment:
are you sure? the `return_data[0]` is actually filled with the `ret_offset` but the returned pointer is `return_data[1]`, so this first value is "hidden" from the the outer world

Reply:
yeah, you are right. late night misunderstanding on my end. morning mind sees this as more straightforward than i thought!